_posts/2025-11-21-verilator-extension.md

```diff
@@ -0,0 +1,207 @@
+Let's begin with a simple counter module:
+
+```verilog
+module counter (
+    input logic clk,
+    input logic reset,
+    input logic en,
+    output logic [31:0] counter_out
+);
+
+    logic [31:0] counter_reg;
+
+    always_ff @(posedge clk or posedge reset) begin
+        if (reset)
+            counter_reg <= 32'd0;
+        else if (en)
+            counter_reg <= counter_reg + 32'd1;
+    end
+
+    assign counter_out = counter_reg;
+
+endmodule
+```
+
+Next, we define a top module that instantiates several controllers, each of which instantiates the counter module. This lets the example demonstrate instance hierarchies, while keeping the complexity as low as possible:
+
+```verilog
+module top (
+    input wire clk,
+    input wire reset,
+    output wire [31:0] counter1,
+    output wire [31:0] counter2,
+    output wire [31:0] counter3,
+    output wire [31:0] counter4,
+    output wire [31:0] counter5
+);
+
+    controller uut(clk, reset, counter1);
+    controller uut1(clk, reset, counter2);
+    controller uut2(clk, reset, counter3);
+    controller uut3(clk, reset, counter4);
+    controller uut4(clk, reset, counter5);
+
+endmodule
+
+module controller(
+    input wire clk,
+    input wire enable,
+    input wire reset,
+    output wire [31:0] counter
+);
+
+    counter cut(clk, reset, counter);
+
+endmodule
+```
+
+With the design in place, the next step is to decide which signals to target for hook insertion.
+The feature supports inserting hooks on multiple signals by adding multiple configuration lines.
+
+Since we've reached the configuration step, let's show how to configure hook insertion for a specific signal.  
+The feature extends Verilator's configuration format (file typically `.vlt`) with entries that describe hook insertion. Each entry specifies:
+- the target signal path
+- the C/C++ callback function name
+- an ID used to select different fault behaviors inside the callback
+
+An expected configuration line for the _hook-insertion_ looks like:  
+`insert_hook -callback "<c-function name>" -id <fault-id> -target "<topmodule.instance.instance.instance.var>"`  
+Let's go through this configuration line and elaborate what each flag needs as input, using our example.
+
+Starting with the `insert_hook` flag, this flag represents the keyword to enable the _hook-insertion_ and therefore does not need an input.
+
+Continuing with the `-target` flag, which is probably the more complicated flag in this configuration. Let's define the amount and which signals we want to _hook-insert_ as stated earlier. For this example we want to _hook-insert_ one signal, which will be the `counter_reg` signal. Since the `-target` flag needs the path to the variable as input, this "target"-path should start with the top module followed by the instances calling the module containing the targeted signal.  
```

Comment:
```suggestion
Continuing with the `-target` flag, which is probably the more complicated flag in this configuration. It defines the amount and signal we want to _hook-insert_ as stated earlier. For this example we want to _hook-insert_ one signal, which will be the `counter_reg` signal. Since the `-target` flag needs the path to the variable as input, this "target"-path should start with the top module followed by the instances calling the module containing the targeted signal.  
```
maybe